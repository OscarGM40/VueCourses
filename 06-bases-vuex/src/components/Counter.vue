<template>
  <h1>Counter - Vuex</h1>
  <h2>Direct access: {{ $store.state.counter.count }}</h2>
  <h2>Computed {{ countComputed }}</h2>
  <h2>ShitComputed {{ anotherShitComputed }}</h2>
  <h1>Map State</h1>
  <h2>mapState: {{ counter }}</h2>
  <h2>lastMutation: {{ lastMutation }}</h2>
  <h2>Direct getter:{{ $store.getters['counter/squareCount']}} </h2>

<button @click="increment" :disabled="isLoading" >Increment +1</button>
<button @click="incrementBy" :disabled="isLoading">Increment +5</button>
<button @click="randomInt" :disabled="isLoading">Random</button>
</template>

<script>
import { mapState, mapActions, mapGetters } from "vuex";

// console.log(mapState('counter'["count","lastMutation","isLoading"]));
// console.log(mapActions(["incrementRandomInt"]));

export default {
  methods:{
    increment(){
      this.$store.commit('counter/increment'); // ojo es el nombre de la mutación
    },
    incrementBy(){
      this.$store.commit('counter/incrementBy',5); // ojo es el nombre de la mutación(con el modulo prefijado)
    },
/*     incrementRandomInt(){
      this.$store.dispatch('incrementRandomInt')
    }, */
    // forma Uno
    // ...mapActions(['incrementRandomInt'])
    // forma Dos que permite renombrar
    ...mapActions({
      randomInt: 'counter/incrementRandomInt'
    })
  },
  computed: {
    countComputed() {
      return this.$store.state.counter.count;
    },
    anotherShitComputed() {
      return this.countComputed;
    },
    ...mapState({
      counter: state => state.counter.count,
      lastMutation: state => state.counter.lastMutation,
      isLoading: state => state.counter.isLoading
    }),
  },
};
</script>

<style></style>
