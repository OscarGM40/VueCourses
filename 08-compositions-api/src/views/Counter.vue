<template>
  <h1>Counter View</h1>
  <h2>{{ counter }}</h2>
  <button @click="increment">+1</button>
  <button @click="decrement">-1</button>
</template>
<script>
import { useCounter } from "@/composables/useCounter";
export default {
  name: "counter",
  props: {},
  emits: [],

  setup() {
    // fijate que usar un composable es lo mejor,ya que abstraigo la lógica y la dejo reusable(dos por uno)
    const { counter, increment, decrement } = useCounter(5);
    return { counter, increment, decrement };

    // const counter = ref(5);
    
    // FORMA UNO declarar las funciones aqui
    /*     const increment = () => {
      return counter.value++;
    };

    const decrement = () => {
      return counter.value--;
    }; */
    // FORMA DOS aprovechar que el return es un objeto para no definir las funciones(no creo que sea muy útil en la vida real esto)
    /*     return {
      counter,
      increment: () => counter.value++,
      decrement: () => counter.value--,
    }; */
  },
};
</script>
<style></style>
