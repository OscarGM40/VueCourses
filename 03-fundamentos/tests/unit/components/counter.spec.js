import { shallowMount } from "@vue/test-utils";
// en Vue la @ hace referencia al punto de entrada,es buena idea usarlo
import Counter from "@/components/Counter";

describe("Counter Component: ", () => {
  test("should match with its snapshot", () => {
    // una prueba interesante puede ser comprobar el snapshot ??

    //  los componentes en un ambiente de testing se renderizan en un DOM virtual,de alguna manera tenemos que montarlos en ese DOM.Aqui es donde entran métodos como shallowMount o mount(éste monta todo el componente y sus componentes,luego el proceso es más pesado)
    // obviamente de momento nos vale shallowMount
    const wrapper = shallowMount(Counter);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('h2 tag should render default value(Counter) ', () => {
    // de nuevo no necesitamos una copia profunda
    const wrapper = shallowMount(Counter);

    // puedo asegurarme que el DOMWrapper existe antes de aplicarle .text()
    expect(wrapper.find("h2").exists()).toBeTruthy();

    // find devuelve un DOMWrapper,para ver el contenido tengo que usar la función DOMWrapper.text()
    const h2 = wrapper.find("h2");
    console.log("first h2 content: ", h2.text());

    expect(h2.text()).toBe("Counter");
  })

  test('p tag should have the default value(10) ', () => { 
    const wrapper = shallowMount(Counter);
    // findAll devuelve un arreglo de DOMWrappers,pero es muy mala idea buscar en el DOM por posiciones,es algo muy volátil esto.Se recomienda usar data-attributes(y removerlos con el plugin de babel en producción)

    // const pTags = wrapper.findAll('p'); <- mala idea buscar por posiciones,asinto
    // ojo que .text() devuelve el contenido como string(aunque fuera un Number en el component)
    // expect(pTags[1].text()).toBe("10");

    // fijate que sin usar plugins con jest puedo buscar usando wrapper.find('[data-testid="value"]').Puedo buscar por cualquier data-attribute,ojo
    const pByDataAttribute = wrapper.find('[data-testid="counter"]')
    expect(pByDataAttribute.text()).toBe('10')
   })
});
