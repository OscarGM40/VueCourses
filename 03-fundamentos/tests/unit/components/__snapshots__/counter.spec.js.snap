// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component:  should match with its snapshot 1`] = `
<h2>Counter</h2>
<p>10 <sup>2</sup> = 100</p>
<p data-testid="counter">10</p>
<div><button type="submit">+1</button><button type="submit">-1</button></div>
`;
